value-processor/src/org/immutables/value/processor/meta/AttributeBuilderReflection.java

```diff
@@ -0,0 +1,434 @@
+package org.immutables.value.processor.meta;
+
+import com.google.common.base.Optional;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import javax.lang.model.element.Modifier;
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeKind;
+import javax.lang.model.type.TypeMirror;
+import org.immutables.value.Value.Immutable;
+import org.immutables.value.Value.Lazy;
+import org.immutables.value.Value.Parameter;
+import org.immutables.value.Value.Style;
+import org.immutables.value.Value.Style.ImplementationVisibility;
+import org.immutables.value.processor.meta.AttributeBuilderDescriptor.ValueToBuilderTarget;
+
+/**
+ * Reflects over the returnType and generates strings for the {@link AttributeBuilderDescriptor}
+ */
+@Immutable(builder = false)
+@Style(visibility = ImplementationVisibility.PRIVATE)
+public abstract class AttributeBuilderReflection {
+
+  static Map<TypeMirror, AttributeBuilderReflection> analyzedReturnTypes = new HashMap<>();
+
+  //TODO: not 100% cached because returnType can be a list (or map etc.)
+  // Maybe bad to cache? Some safe way to cache without holding onto processing environment?
+  public static AttributeBuilderReflection forValueType(ValueAttribute valueAttribute) {
```

Comment:
the caching here may be dangerous, saw another comment that ValueAttribute has a handle to the environment which can cause memory leaks. Better way to cache?